### PR TITLE
imodel transformer struct instance copy tests

### PR DESCRIPTION
**Pull request by MichaelBelousov**
- add tests for copying elements with structs and struct instances to test new native behavior
- refactor `processSchemas` to be based on `exportSchemas`
- make `onExportSchema` async and `shouldExportSchema` take a schema key, not a full parsed schema object
- added (already existing in monorepo) dependency on semver to imodel-transformer
- [node addon PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/164739)

### CODE REVIEW

File: test-apps/imodel-transformer/src/Transformer.ts
Comment (MichaelBelousov) [inline test-apps/imodel-transformer/src/Transformer.ts:212]: this was intentionally kept from previous debugging, will remove

**Comment by MichaelBelousov** [comment]
converting to draft in case the schema editing method changes

**Review by mgooding — APPROVED** [review]

File: core/backend/package.json
Comment (MichaelBelousov) [inline core/backend/package.json:98]: reverting

File: core/backend/src/IModelExporter.ts
Comment (MichaelBelousov) [inline core/backend/src/IModelExporter.ts:120]: @scsewall I initially tried to keep this from being async, but the code was much worse. I can imagine paint points unless most of these handlers become async, thoughts?

**Comment by MichaelBelousov** [comment]
the StructArray test crashes with the current code, there will be an imodel02 PR soon to fix it

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [reply core/backend/src/IModelExporter.ts:120]: Most of the `onExport*` callbacks are handling a single entity, so I am not sure that making them async is required. However, I can see the merit in making `onExportSchema` async.

File: test-apps/imodel-transformer/package.json
Comment (scsewall) [inline test-apps/imodel-transformer/package.json:30]: Minor nit - I like to alphabetize dependencies. If you don't, then you have to read the entire list when looking for a particular dependency.

Comment (MichaelBelousov) [reply test-apps/imodel-transformer/package.json:30]: yeah I used `rush add` to add this dep and iirc it doesn't alphabetize, will fix

**Comment by MichaelBelousov** [comment]
rebased on imodel02 and retargeted imodel02

**Comment by MichaelBelousov** [comment]
it looks like regex's weaknesses surfaced immediately. Upon rerunning with the imodel02 branch, I noticed serialization doesn't preserve any element or attribute order (which is fine ofc), but that's not worth using in regex which can only "parse" xml if there is a guaranteed order. While it's possible to quickly prototype an `xslt` schema changing backend since schemas are in xml, probably I'll just add a callback option to the test-app transformer for tests to modify the schema's xml dom directly in js

**Comment by MichaelBelousov** [comment]
@scsewall please let me know what you think of how I decided to change `processSchemas` with the same behavior while still adding a default implementation of `exportSchema` that does full xml parsing, and sharing some logic with `shouldExportSchema`

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [inline core/backend/src/IModelExporter.ts:120]: Please mention this API change in NextVersion.md

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [inline core/backend/src/IModelExporter.ts:320]: Because of dependencies, the order that schemas are exported to the target must match the order that they were imported into the source. Is the above Promise.all code guaranteed to maintain that order?

File: core/backend/src/IModelTransformer.ts
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:34]: Looks like you need to sort imports.

File: core/backend/package.json
Comment (scsewall) [inline core/backend/package.json:135]: Why is xmldom needed?

File: core/backend/src/IModelTransformer.ts
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:757]: Doesn't the call to querySchemaVersion on the target eliminate the need for this method?

File: core/backend/src/IModelTransformer.ts
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:799]: It is more efficient to import all schemas at once rather than doing it one by one.

File: core/backend/src/IModelTransformer.ts
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:795]: Is there a way to just write the schema XML file out without parsing it? Core code produced it so it should be valid plus the XML will be validated in native code before importing.

**Review by scsewall — CHANGES_REQUESTED** [review]

File: core/backend/src/IModelExporter.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts:320]: good point, it is not guaranteed. A more complex barrier could do it and keep the parallelization, but I think I'll just remove the Promise.all.

File: core/backend/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:34]: could have sworn `rush lint` didn't pick anything up... will try again

File: core/backend/package.json
Comment (MichaelBelousov) [reply core/backend/package.json:135]: `core/backend/src/IModelTransformer.ts` uses it for the default `onExportSchema` implementation which serializes the `ECSchemaMetaData.Schema` object. I couldn't find any other way to serialize that object. Or a way to export specific schemas from nativeDb. Is there any other example that shows how `onExportSchema` was intended to be overrided if the xml is needed to be read?

File: core/backend/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:757]: this method is for the filter on schema paths that are exported by `nativeDb.exportSchemas` in `processSchemas`. If we can't rely on a naming convention for the exported files, then we need to read the file for the schema name and version to know what we're importing from the db. We could query the sourceDb for its schema versions but I don't think we can correlate it to the exported files without doing the same amount of work.

File: core/backend/package.json
Comment (scsewall) [reply core/backend/package.json:135]: Maybe this is a change request for the `ecschema-metadata` package then. Seems like there should be a `toString` or `writeToFile` method on the `Schema` class. Seems inefficient to create an XMLDom just to write out a file.

File: core/backend/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:799]: my thoughts on this are that `processSchemas` is the efficient batch method. This if for people that want a more granular per-schema export functionality. I think a better architecture would be something like:
```ts
public async onReadyExportSchema(schema: Schema) {
}

public async onExportSchemas() {
  const readyTransformedSchemas = await Promise.all(getSchemas().map(onReadyExportSchema));
  // so schemas get "ready" in parallel but are imported efficiently in one call to this.targetDb.importSchemas after the barrier
  batchImportSchemas(readyTransformedSchemas);
}
```

File: core/backend/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:795]: that isn't actually parsing the schema. It's parsing just the xml header string: `<?xml version="1.0" encoding="UTF-8"?>`
the `schema.toXml` then mutates that document object, without any parsing. Maybe I don't even need to parse the xml-header but it seemed the best way to make sure the document contains the xml header after being filled.

Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:795]: going to replace with an added alpha API method `ECSchemaMetaData.Schema.toXmlString`

**Comment by pmconne** [comment]
Why has the base branch been switched back to master? The tests can't pass without the corresponding native changes, right?

**Comment by MichaelBelousov** [comment]
> Why has the base branch been switched back to master? The tests can't pass without the corresponding native changes, right?

Was discussing with @scsewall and decided to wait for the addon release that picks up the native changes... 
Also rebasing on master appears to have added many many code owners and I'm sorry about that.

**Comment by pmconne** [comment]
> Was discussing with @scsewall and decided to wait for the addon release that picks up the native changes...

Ok, though it'd be preferable for the native changes and the corresponding tests to be merged into master together. Are you producing the new addon?

> Also rebasing on master appears to have requested reviews from many many code owners and I'm sorry about that.

Only the five people with a shield symbol next to their names are actually required to review.

**Comment by MichaelBelousov** [comment]
> Are you producing the new addon?

I can produce a new addon when this PR is ready. If we shouldn't have switched from basing on imodel02 I can revert. I haven't checked if the native PR made it into the addon.

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [reply core/backend/src/IModelExporter.ts:328]: Were you going to refactor to not use Promise.all?

File: core/backend/src/IModelTransformer.ts
Comment (scsewall) [inline core/backend/src/IModelTransformer.ts:755]: Is shouldExportSchemaCheck called from anywhere else? If not, it could be combined into shouldExportSchema.

File: core/backend/src/test/IModelTransformerUtils.ts
Comment (scsewall) [inline core/backend/src/test/IModelTransformerUtils.ts:1655]: Our convention is to use `return` instead of `await` for the last line of a method.

File: core/backend/src/IModelTransformer.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelTransformer.ts:755]: you're right, the second call path was removed, should be able to remove that now

File: core/backend/src/IModelExporter.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts:328]: it does run `onExportSchema` using `Promise.all` here, was there somewhere else you were referring to?

File: docs/changehistory/NextVersion.md
Comment (scsewall) [inline docs/changehistory/NextVersion.md:62]: I would rather encourage a quick answer in `shouldExportSchema` (don't load the schema) and then skip over it in `onExportSchema` if you need it to be loaded to get your answer.

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [reply core/backend/src/IModelExporter.ts:328]: Schema order is important and I thought Promise.all was not guaranteed to have a deterministic order? Instead you could iterate the `schemasToExport` array and await each call to `onExportSchema`.

Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts:328]: I thought schema order was only important for the actual import, and the import should still happens in the same order as it did before this PR even with the `Promise.all`.
Though now that I think about it, it only happens in whatever order the implementation of `readdirSync` uses which is platform dependent. Afaict, the previous implementation of `processSchemas` did not order the imports differently.
Am I missing something?

Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts:328]: If the `meta.ECSchemaDef` query is in schema topological/reference order then that's probably it. Yes this may export them out of order. I suppose I will just remove any parallelism for the time being.

Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts]: I also now realize you wrote "refactor to *not* use Promise.all", I somehow missed the "not" (even though the comment that way is contextually pointless). Either way fixed.

File: core/backend/src/IModelExporter.ts
Comment (MichaelBelousov) [inline core/backend/src/IModelExporter.ts:317]: wrong, fixing

File: core/backend/src/IModelExporter.ts
Comment (scsewall) [reply core/backend/src/IModelExporter.ts]: Sorry, you can probably keep things as you have them. Schema order is important if you import them one by one. Now that you are importing them as a collection after all exports, then this should be okay since importSchemas doesn't care about the order - it just has to have dependencies in the list.

**Comment by MichaelBelousov** [comment]
@pmconne 

> I expect @markschlosseratbentley will want a new addon incorporating his changes-in-progress before 2.17.0 is released, so we'll probably want to wait until those are ready.
Can you confirm that the tests added by this PR definitely pass using an addon built from imodel02 master?

They definitely did on the branch that definitely merged into master, I will pull master and double check to confirm.

File: core/backend/src/IModelExporter.ts
Comment (MichaelBelousov) [reply core/backend/src/IModelExporter.ts]: I did forget that passing them as a list lets the native db handle sorting them before import. Alright well in that case I'll revert my revert etc and keep exporting schemas parallel

**Comment by pmconne** [comment]
> They definitely did on the branch that definitely merged into master, I will pull master and double check to confirm.

No need - thanks for confirming.

**Review by scsewall — APPROVED** [review]

**Pull request merged**
